### DISCUSSION TITLE
Set ntp time partly working

### DISCUSSION THREAD

**Discussion by Radulfur** [dewenni/ESP_Buderus_KM271#16] [Q&A]
Hi there,
I‘m using this software slightly extended in order to attempt some BFU integration.
Setting ntp time seems to work fine at first glance. The WebGUI shows time correctly but the display which is part of my HS2105 shows some completely different time. 
Are there any ideas why this might happen?

Best regards

**Comment by dewenni** [comment]
Hi,

are you using the last version of the software? Because I did some bugfixing regarding the date & Time function at the Logamatic.
The final tests with my R2107 looked good for me.

But maybe there is a difference to the HS2105?!

Next thing to mention is, that the NTP Time on the ESP, that is also shown in the WebUI is not strictly the same as the time at the Logamatic.
The NTP time from the ESP is used to write the date & time information to the Logamatic, but I don´t know how to read the date & time information from the Logamatic back to the ESP.

There are two functions to write the date & time informations to the Logamatic:
`km271SetDateTimeDTI(tm dti)` => write Date&Time Informations entered from WebUI to the Logamatic
`km271SetDateTimeNTP()` => write Date&Time Informations based on NTP from ESP to the Logamatic

**Comment by Radulfur** [comment]
Hi,
this is strange. I performed my check again and it works now. The only difference was that I tested it first directly after booting the HS2105. The time displayed on my HS2105 was 00:00 then, which means that setting the new time would have meant a great step forward (or backward) in time. In my new test there was only a difference of a few minutes.
My HS2105 seems to drift a lot in time. Unfortunately there's no option to adjust this drift like it could be done on the R2107.

Thanks a lot for your work and assistance!

**Comment by dewenni** [comment]
Hi,
it takes some time after bootup to get the ntp date and time from the NTP-Server.
So if you set the ntp time just after bootup, it could be that 01.01.1970 00:00 will be written to the Logamatic.

I have also implemented, that every hour the DST-Bit (daylight saving time) of the NTP information has changed.
If so, the Date&Time on the Logamatic will be written automatically.

```
// check every hour if DST has changed
  #ifdef USE_NTP
    if (dstTimer.cycleTrigger(3600000))
    {
      time_t now;                       
      tm dti;                           
      time(&now);                                   // read the current time
      localtime_r(&now, &dti);                      // update the structure tm with the current time
      if (dst_old!=dti.tm_isdst && !main_reboot){   // check if DST has changed
        km271SetDateTimeNTP();                      // change date and time on buderus
      }
      dst_old=dti.tm_isdst;    
    }
  #endif
```

You could also think about to write the Date&Time every hour to the Logamatic independent of the DST-Bit.
like this:

```
// correct Date&Time every hour
  #ifdef USE_NTP
    if (dstTimer.cycleTrigger(3600000))
    {
      km271SetDateTimeNTP();                      // change date and time on buderus
    }
  #endif
```

**Comment by dewenni** [comment]
Can you also tell us something about this:

> I‘m using this software slightly extended in order to attempt some BFU integration.

**Comment by Radulfur** [comment]
Yes, I already thought about adjusting time somehow regularly if the difference between HS2105 time and NTP becomes greater than let's say 5 minutes. As there is no known solution to request time information from the HS2105, this won't work. Maybe I will go for setting the time once a week or month. That should be sufficient.

My attempted BFU integration is somehow primitive. The idea behind this is to replace the optional external sensor of the BFU by a digital potentiometer - a DS3502 in my case - and to send room temperature information to the ESP32 via WiFi and MQTT. 
The digipot can be controlled via I2C. Therefore I soldered a JST connector to the EEPROM slot (older board from Daniel/the78mole). 
It's quite easy to extend your code in order to control the DS3502. However I'm still fighting with the physical connection to the BFU. It works in general but the bandwidth which can be set is still limited to about 17 to 22°C. According to the NTC characteristics documented in the service manual a 10k digipot should be sufficient to set room temperature between 8 to 25°C but it's not - at least not as I connected it (I'm not an electronics engineer and I don't have any infomation on the BFU schematics). 
I'll still have to investigate but I'll be happy to inform you if you like to. We should probably open a separate discussion then, right?

**Comment by Radulfur** [comment]
Cabling of DS3502 was not as it should be. Bandwidth is now between 14°C and 34.5°C.
That should be sufficient for my purposes.
